src/sqlfluff/utils/reflow/depthmap.py

```diff
@@ -12,16 +12,36 @@
 reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
 
 
-def _stack_pos_interpreter(path_step: PathStep):
-    """Interpret a path step for stack_positions."""
-    if path_step.idx == 0 and path_step.idx == path_step.len - 1:
-        return "solo"
-    elif path_step.idx == 0:
-        return "start"
-    elif path_step.idx == path_step.len - 1:
-        return "end"
-    else:
-        return ""  # NOTE: Empty string evaluates is falsy.
+@dataclass(frozen=True)
+class StackPosition:
+    """An element of the stack_positions property of DepthInfo."""
+
+    idx: int
+    len: int
+    type: str
+
+    @staticmethod
+    def _stack_pos_interpreter(path_step: PathStep):
+        """Interpret a path step for stack_positions."""
+        if path_step.idx == 0 and path_step.idx == path_step.len - 1:
+            return "solo"
+        elif path_step.idx == 0:
+            return "start"
+        elif path_step.idx == path_step.len - 1:
+            return "end"
+        else:
+            return ""  # NOTE: Empty string evaluates is falsy.
+
+    @classmethod
+    def from_path_step(cls, path_step: PathStep):
```

Comment:
That formulation on it's own doesn't work, I'll change it to: `def from_path_step(cls: Type["StackPosition"], path_step: PathStep) -> "StackPosition":` which does. Otherwise it errors that `T` is not defined.

Reply:
I think this is it.

https://twitter.com/willmcgugan/status/1397636465404915719?s=20&t=INxWfvN8AwLA8V7RwAKTjA

Reply:
Yeah that works - it needs the `TypeVar` definition at the top of the file. I think being explicit here is overall simpler.